### PR TITLE
Homework mock

### PR DESCRIPTION
**Pull request by Kuba-Wi**

### CODE REVIEW

File: src/ThreadTimer.hpp [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+#include <thread>
+#include "Timer.hpp"
+
+class ThreadTimer : public Timer {
+public:
+    virtual void sleep_for(minutes duration) override{
```
Comment (Kasperice) [inline src/ThreadTimer.hpp:7]: Is this 'virtual' needed here? (I am just asking)

**Review by Kasperice — APPROVED** [review]
Good code, I had one question below

File: src/ThreadTimer.hpp (hunk 1 above)
Comment (Kuba-Wi) [reply src/ThreadTimer.hpp:7]: It is not

**Review by WBigaj — APPROVED** [review]
I have checked your code and it looks solid.

File: test/PizzeriaTest.cpp [hunk 2]
```diff
@@ -65,3 +63,23 @@ TEST_F(PizzeriaTest, calculatePriceForPizzaMock)
 
     delete mock;
 }
+
+TEST_F(PizzeriaTest, orderStubAndTwoMockPizzas) {
+    // Given
+    StrictMock<PizzaMock>* strict_mock = new StrictMock<PizzaMock>{"strict", 1, minutes(0)};
+    NiceMock<PizzaMock>* nice_mock = new NiceMock<PizzaMock>{"nice", 2, minutes(3)};
+    Pizzas pizzas = {new PizzaStub{"STUB"}, strict_mock, nice_mock};
```
Comment (jakjan95) [inline test/PizzeriaTest.cpp:71]: I think that here will be better create Mocks as a objects and passing them into pizzas by reference :)

**Review by jakjan95 — APPROVED** [review]
Generally looks good :)

File: test/PizzeriaTest.cpp (hunk 2 above)
Comment (Kuba-Wi) [reply test/PizzeriaTest.cpp:71]: Yeah, definitely it would be easier

File: test/PizzeriaTest.cpp [hunk 3]
```diff
@@ -65,3 +63,23 @@ TEST_F(PizzeriaTest, calculatePriceForPizzaMock)
...
+    EXPECT_CALL(tm, sleep_for(_)).Times(3);
```
Comment (ziobron) [inline test/PizzeriaTest.cpp:72]: _ means that it doesn't matter what argument is passed. We would like to check if sleep_for was called with the baking time of each pizza. In your case, it is not tested.

File: test/PizzeriaTest.cpp [hunk 4]
```diff
@@ -65,3 +63,23 @@ TEST_F(PizzeriaTest, calculatePriceForPizzaMock)
...
+    EXPECT_CALL(*strict_mock, getPrice());
+    EXPECT_CALL(*strict_mock, getName());
+    EXPECT_CALL(*strict_mock, getBakingTime());
+
+    // When
+    auto orderId = pizzeria.makeOrder(pizzas);
+    auto price = pizzeria.calculatePrice(orderId);
```
Comment (ziobron) [inline test/PizzeriaTest.cpp:79]: what about that price? It would be nice to check if it is equal to expected price.

**Review by ziobron — CHANGES_REQUESTED** [review]

File: test/PizzeriaTest.cpp [hunk 5]
```diff
@@ -9,6 +9,11 @@
 using namespace std;
 using namespace ::testing;
 
+constexpr double niceMockPizzaPrice = 35.0;
+constexpr int niceMockPizzaBakingTime = 5;
+constexpr double strictMockPizzaPrice = 25.0;
+constexpr int strictMockPizzaBakingTime = 8;
```
Comment (ziobron) [inline test/PizzeriaTest.cpp:15]: Why you are not using those values?

File: test/PizzeriaTest.cpp [hunk 6]
```diff
@@ -66,20 +71,25 @@ TEST_F(PizzeriaTest, calculatePriceForPizzaMock) {
...
     NiceMock<PizzaMock>* nice_mock = new NiceMock<PizzaMock>{"nice", 2, minutes(3)};
-    Pizzas pizzas = {new PizzaStub{"STUB"}, strict_mock, nice_mock};
-    EXPECT_CALL(tm, sleep_for(_)).Times(3);
-    EXPECT_CALL(*strict_mock, getPrice());
+    PizzaStub ps("STUB");
+    Pizzas pizzas = {&ps, strict_mock, nice_mock};
+    EXPECT_CALL(*strict_mock, getPrice()).WillRepeatedly(Return(10.0));
     EXPECT_CALL(*strict_mock, getName());
     EXPECT_CALL(*strict_mock, getBakingTime());
+    EXPECT_CALL(*nice_mock, getPrice()).WillOnce(Return(10.0));
```
Comment (ziobron) [inline test/PizzeriaTest.cpp:81]:
```suggestion
    EXPECT_CALL(*nice_mock, getPrice()).WillOnce(Return(niceMockPizzaPrice));
```

File: test/PizzeriaTest.cpp [hunk 7]
```diff
@@ -66,20 +71,25 @@ TEST_F(PizzeriaTest, calculatePriceForPizzaMock) {
...
     // Given
-    StrictMock<PizzaMock>* strict_mock = new StrictMock<PizzaMock>{"strict", 1, minutes(0)};
+    StrictMock<PizzaMock>* strict_mock = new StrictMock<PizzaMock>{"strict", 1, minutes(1)};
     NiceMock<PizzaMock>* nice_mock = new NiceMock<PizzaMock>{"nice", 2, minutes(3)};
-    Pizzas pizzas = {new PizzaStub{"STUB"}, strict_mock, nice_mock};
-    EXPECT_CALL(tm, sleep_for(_)).Times(3);
-    EXPECT_CALL(*strict_mock, getPrice());
+    PizzaStub ps("STUB");
+    Pizzas pizzas = {&ps, strict_mock, nice_mock};
+    EXPECT_CALL(*strict_mock, getPrice()).WillRepeatedly(Return(10.0));
```
Comment (ziobron) [inline test/PizzeriaTest.cpp:78]:
```suggestion
    EXPECT_CALL(*strict_mock, getPrice()).WillRepeatedly(Return(strictMockPizzaPrice));
```

File: test/PizzeriaTest.cpp [hunk 8]
```diff
@@ -66,20 +71,25 @@ TEST_F(PizzeriaTest, calculatePriceForPizzaMock) {
...
+    EXPECT_CALL(tm, sleep_for(_)).Times(2);
+    EXPECT_CALL(tm, sleep_for(ps.getBakingTime()));
 
     // When
     auto orderId = pizzeria.makeOrder(pizzas);
     auto price = pizzeria.calculatePrice(orderId);
     pizzeria.bakePizzas(orderId);
     pizzeria.completeOrder(orderId);
 
+    ASSERT_EQ(price, 30);
```
Comment (ziobron) [inline test/PizzeriaTest.cpp:91]:
```suggestion
    ASSERT_EQ(price, strictMockPizzaPrice + niceMockPizzaPrice + ps.getPrice());
```

File: test/PizzeriaTest.cpp [hunk 9]
```diff
@@ -66,20 +71,25 @@ TEST_F(PizzeriaTest, calculatePriceForPizzaMock) {
...
+    StrictMock<PizzaMock>* strict_mock = new StrictMock<PizzaMock>{"strict", 1, minutes(1)};
     NiceMock<PizzaMock>* nice_mock = new NiceMock<PizzaMock>{"nice", 2, minutes(3)};
-    Pizzas pizzas = {new PizzaStub{"STUB"}, strict_mock, nice_mock};
-    EXPECT_CALL(tm, sleep_for(_)).Times(3);
-    EXPECT_CALL(*strict_mock, getPrice());
+    PizzaStub ps("STUB");
+    Pizzas pizzas = {&ps, strict_mock, nice_mock};
+    EXPECT_CALL(*strict_mock, getPrice()).WillRepeatedly(Return(10.0));
     EXPECT_CALL(*strict_mock, getName());
     EXPECT_CALL(*strict_mock, getBakingTime());
```
Comment (ziobron) [inline test/PizzeriaTest.cpp:80]:
```suggestion
    EXPECT_CALL(*strict_mock, getBakingTime()).WillOnce(Return(minutes(strictMockPizzaBakingTime));
```

File: test/PizzeriaTest.cpp [hunk 10]
```diff
@@ -66,20 +71,25 @@ TEST_F(PizzeriaTest, calculatePriceForPizzaMock) {
...
+    EXPECT_CALL(*nice_mock, getPrice()).WillOnce(Return(10.0));
+    EXPECT_CALL(tm, sleep_for(_)).Times(2);
```
Comment (ziobron) [inline test/PizzeriaTest.cpp:82]:
```suggestion
    EXPECT_CALL(tm, sleep_for(minutes(strictMockPizzaBakingTime)));
    EXPECT_CALL(tm, sleep_for(minutes(niceMockPizzaBakingTime)));
```

**Review by ziobron — CHANGES_REQUESTED** [review]

**Comment by Kuba-Wi** [comment]
@ziobron I commited your suggestions and now tests don't pass.

File: test/PizzeriaTest.cpp [hunk 11]
```diff
@@ -65,3 +75,29 @@ TEST_F(PizzeriaTest, calculatePriceForPizzaMock)
...
+TEST_F(PizzeriaTest, orderStubAndTwoMockPizzas) {
+    // Given
+    StrictMock<PizzaMock>* strict_mock = new StrictMock<PizzaMock>{"strict", 1, minutes(1)};
+    NiceMock<PizzaMock>* nice_mock = new NiceMock<PizzaMock>{"nice", 2, minutes(3)};
+    PizzaStub ps("STUB");
+    Pizzas pizzas = {&ps, strict_mock, nice_mock};
+    EXPECT_CALL(*strict_mock, getPrice()).WillRepeatedly(Return(strictMockPizzaPrice));
+    EXPECT_CALL(*strict_mock, getName());
+    EXPECT_CALL(*strict_mock, getBakingTime()).WillOnce(Return(minutes(strictMockPizzaBakingTime)));
+    EXPECT_CALL(*nice_mock, getPrice()).WillOnce(Return(niceMockPizzaPrice));
```
Comment (ziobron) [inline test/PizzeriaTest.cpp:88]: you are missing expect call on nice mock - this one about baking time. This time is expected later by sleep_for below.

**Review by ziobron — CHANGES_REQUESTED** [review]
Yes, one oversight

**Comment by Kuba-Wi** [comment]
Now tests fail because valgrind is not installed.

**Comment by ziobron** [comment]
> @ziobron Now tests fail because valgrind is not installed.

Probably you know what to do :) Install it on CI.
Take a look on others people PRs. You just need to add one line into workflow yml file.

File: .github/workflows/module3.yml [hunk 12]
```diff
@@ -8,6 +8,8 @@ jobs:
     steps:
     - name: Check out code
       uses: actions/checkout@master
+    - name: install valgrind
+      uses: sudo apt-get update && sudo apt-get install valgrind
```
Comment (ziobron) [inline .github/workflows/module3.yml:12]:
```suggestion
      run: sudo apt-get update && sudo apt-get install valgrind
```

**Review by ziobron — APPROVED** [review]
Travis CI works so I have a proof of no leaks. 
To make GitHub Actions with you need to have `run` instead of `uses` here.

**Comment by ziobron** [comment]
{ "requirement_ids" : [1119, 1119, 1119, 3524, 3525, 3526] }
3 * 3 XP za punktualność (1119)
10 XP za TimeMock w Pizzas (3524)
15 XP za nowy przypadek testowy w Pizzas (3525)
5 XP za workflow w Pizzas (3526)

**Pull request closed without merging**